providers/implementations/encode_decode/ml_kem_codecs.c

```diff
@@ -0,0 +1,559 @@
+};
+
+/* Indices of slots in the `cinfo_map` table below */
+#define ML_KEM_512_CINFO    0
+#define ML_KEM_768_CINFO    1
+#define ML_KEM_1024_CINFO   2
+
+/*
+ * Per-variant fixed parameters
+ */
+static const ML_KEM_CINFO cinfo_map[3] = {
+    { &ml_kem_512_spki_info,  ml_kem_512_pkcs8_info },
+    { &ml_kem_768_spki_info,  ml_kem_768_pkcs8_info },
+    { &ml_kem_1024_spki_info, ml_kem_1024_pkcs8_info }
+};
+
+/* Retrieve the parameters of one of the ML-KEM variants */
+static const ML_KEM_CINFO *ml_kem_get_cinfo(int evp_type)
+{
+    switch (evp_type) {
+    case EVP_PKEY_ML_KEM_512:
+        return &cinfo_map[ML_KEM_512_CINFO];
+    case EVP_PKEY_ML_KEM_768:
+        return &cinfo_map[ML_KEM_768_CINFO];
+    case EVP_PKEY_ML_KEM_1024:
+        return &cinfo_map[ML_KEM_1024_CINFO];
+    }
+    return NULL;
+}
+
+static int vp8_pref_cmp(const void *va, const void *vb)
+{
+    const ML_KEM_PKCS8_PREF *a = va;
+    const ML_KEM_PKCS8_PREF *b = vb;
+
+    /*
+     * Zeros sort last, otherwise the sort is in increasing order.
+     *
+     * The preferences are small enough to ensure the comparison is monotone as
+     * required.  Some versions of qsort(3) have been known to crash when the
+     * comparison is not monotone.
+     */
+    if (a->vp8_pref > 0 && b->vp8_pref > 0)
+        return a->vp8_pref - b->vp8_pref;
+    if (a->vp8_pref == 0)
+        return b->vp8_pref;
+    return -a->vp8_pref;
+}
+
+static ML_KEM_PKCS8_PREF *vp8_order(const char *algorithm_name,
+                                    const ML_KEM_PKCS8_INFO *pkcs8_info,
+                                    const char *direction, const char *formats)
+{
+    ML_KEM_PKCS8_PREF *ret;
+    int i, count = 0;
+    const char *fmt = formats, *end;
+    const char *sep = "\t ,";
+
+    if ((ret = OPENSSL_zalloc((NUM_PKCS8_FORMATS + 1) * sizeof(*ret))) == NULL)
+        return NULL;
+    for (i = 0; i < NUM_PKCS8_FORMATS; ++i) {
+        ret[i].vp8_entry = &pkcs8_info[i];
+        ret[i].vp8_pref = 0;
+    }
+
+    /* Default to compile-time table order. */
+    if (formats == NULL)
+        return ret;
+
+    /* Formats are case-insensitive, separated by spaces, tabs and/or commas */
+    do {
+        if (*(fmt += strspn(fmt, sep)) == '\0')
+            break;
+        end = fmt + strcspn(fmt, sep);
+        for (i = 0; i < NUM_PKCS8_FORMATS; ++i) {
+            if (ret[i].vp8_pref > 0
+                || OPENSSL_strncasecmp(ret[i].vp8_entry->p8_name,
+                                       fmt, (end - fmt)) != 0)
+                continue;
+            ret[i].vp8_pref = ++count;
+            break;
+        }
+        fmt = end;
+    } while (count < NUM_PKCS8_FORMATS);
+
+    if (count == 0) {
+        OPENSSL_free(ret);
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_ML_KEM_NO_FORMAT,
+                       "no %s private key %s formats are enabled",
+                       algorithm_name, direction);
+        return NULL;
+    }
```

Comment:
There should also be an error raised if `count >= NUM_PKCS8_FORMATS`.
I.e if the do while loop exits normally.

Reply:
That's not possible.

Reply:
More to the point once that count reaches the value, we've assigned an ordinal to every entry in the table, and any additional elements in the user's list are redundant. Further iterations of the loop would be non-productive, and harmless. The loop terminating normally is OK.

Reply:
If a user repeats an entry, overflow is possible unless I'm missing something.  Moreover, silently ignoring additional items in the user's list is rather bad form.  Basic API design would mandate a check and error for this condition.  That's all I'm asking for.  You check for zero and error, you should also check for too many to process and error.

Reply:
No overflow is possible.

Reply:
If a user supplies a list with duplicates, the latter entries will be ignored silently.  This is unexpected behaviour.
An error return alerts the user to their mistake.  Silently continuing is poor design.

Reply:
Resolving this was very rude.  Please desist doing so if a disagreement remains.

All I'm wanting is:
```suggestion
    }
    if (count >= NUM_PKCS8_FORMATS) {
        OPENSSL_free(ret);
        ERR_raise(ERR_LIB_PROV, PROV_R_LENGTH_TOO_LARGE);
        return NULL;
    }
```

My trusty HP 55 says this took me under two minutes.
Was that so difficult?

Reply:
@openssl/committers second opinion?